### PR TITLE
Add a note that gplate is not uniplate

### LINKED ISSUE #966: Control.Lens.rewriteM does not behave like Data.Generics.Uniplate.Data.rewriteM
Is the intention that Lens version of rewriteM should behave the same as Uniplate version?
The result from the Lens version still has places where the rewrite can be applied, whereas Uniplate version does all the rewrites.

If they are intended to be the same, I can make a proper bug report with an example.

### PR DESCRIPTION
**Pull request by phadej**
Closes #966 

@RyanGlScott do you think we need to add something here? For me it was obvious that `GHC.Generics` based approach cannot handle mutual recursive types (without very ugly tricks, in a way `gplate` would need). Yet it was a surprise to @augustss.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
Ultimately, I think @augustss is the only one who can say whether this is sufficient documentation. From my perspective, however, it may be helpful to include a small example that demonstrates where `gplate`'s treatment of recursive data types differs from `uniplate`. Perhaps a condensed version of the example in https://github.com/ekmett/lens/issues/966#issuecomment-760489484 would suffice.

**Comment by phadej** [comment]
@RyanGlScott, I added a small example.

File: src/Control/Lens/Plated.hs [hunk 1]
```diff
@@ -721,6 +728,41 @@ parts = partsOf plate
...
 
 -- | Implement 'plate' operation for a type using its 'Generic' instance.
+--
+-- Note: the behavior may be different than with 'uniplate' in some special cases.
+-- The 'gplate' doesn't look through other types in a group of mutually
+-- recusive types.
+--
+-- For example consider mutually recursive even and odd natural numbers:
+--
+-- >>> data Even = Z | E Odd deriving (Show, Generic, Typeable, Data); data Odd = O Even deriving (Show, Generic, Typeable, Data)
```
Comment (phadej) [inline src/Control/Lens/Plated.hs:738]: For some reason GHC-7.8 doesn't like multiple line definition.

File: src/Control/Lens/Plated.hs [hunk 2]
```diff
@@ -721,6 +728,41 @@ parts = partsOf plate
 -------------------------------------------------------------------------------
 
 -- | Implement 'plate' operation for a type using its 'Generic' instance.
+--
+-- Note: the behavior may be different than with 'uniplate' in some special cases.
+-- The 'gplate' doesn't look through other types in a group of mutually
```
Comment (RyanGlScott) [inline src/Control/Lens/Plated.hs:733]:
```suggestion
-- 'gplate' doesn't look through other types in a group of mutually
```

File: src/Control/Lens/Plated.hs [hunk 3]
```diff
@@ -721,6 +728,41 @@ parts = partsOf plate
...
+-- recusive types.
```
Comment (RyanGlScott) [inline src/Control/Lens/Plated.hs:734]:
```suggestion
-- recursive types.
```

File: src/Control/Lens/Plated.hs [hunk 4]
```diff
@@ -721,6 +728,41 @@ parts = partsOf plate
...
+--
+-- For example consider mutually recursive even and odd natural numbers:
+--
+-- >>> data Even = Z | E Odd deriving (Show, Generic, Typeable, Data); data Odd = O Even deriving (Show, Generic, Typeable, Data)
+--
+-- Then 'Data' based 'uniplate' finds
```
Comment (RyanGlScott) [inline src/Control/Lens/Plated.hs:740]:
```suggestion
-- Then 'uniplate', which is based on `Data`, finds
```

**Review by RyanGlScott — COMMENTED** [review]
Other than some minor grammar-related suggestions, LGTM.

**Pull request merged**